### PR TITLE
[api] Fix nullptr deref when client teardown reenters state dispatch

### LINKED ISSUE #16798: nullptr dereferenced in api_server.cpp resulting in crash.
### The problem

When the device is restarted via the web service it results in a crash due to a null pointer being dereferenced.
The code in question is:
```
#ifdef USE_LIGHT
API_DISPATCH_UPDATE(light::LightState, light)
#endif
```
This gets expanded to:
```
void APIServer::on_light_update(light::LightState *obj) {
    if (obj->is_internal()) 
      return;
    for (auto &c : this->active_clients()) {
      if (c->flags_.state_subscription) {
        c->send_light_state(obj);
      }
    }
}
```
The `c` is a `nullptr`. I have confirmed this by modifying the `api_server.cpp` and expanding and modifying the function so that it now looks like this:
```
void APIServer::on_light_update(light::LightState *obj) {
    if (obj->is_internal()) 
      return;
    for (auto &c : this->active_clients()) {
      if( c == nullptr ) {
        ESP_LOGD("api", "Client is NULL!");
      }
      if (c->flags_.state_subscription) {
        c->send_light_state(obj);
      }
    }
}
```
The "Client is NULL!" then appears in the logs immediately prior to the crash.
Here is the relevant log entries from a crash without `api_server.cpp` being modified. I've also appended the results of `addr2line` on the backtrace.

```
[I][safe_mode:142]: Boot seems successful; resetting boot loop counter
[D][preferences:148]: Writing 4 items: 3 cached, 1 written, 0 failed
[I][restart.button:011]: Restarting device
[I][app:271]: Rebooting safely
[D][debug:057]: Storing reboot source: web_server
[D][preferences:148]: Writing 2 items: 1 cached, 1 written, 0 failed
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x4200891a  PS      : 0x00060b30  A0      : 0x82008937  A1      : 0x3fca23c0
A2      : 0x00000040  A3      : 0x3fca6d34  A4      : 0x3c145798  A5      : 0x4201df94
A6      : 0x3fca72d8  A7      : 0x3fca72d4  A8      : 0x82008925  A9      : 0x3fca23a0
A10     : 0x00000000  A11     : 0x3fca6d34  A12     : 0x3fca0040  A13     : 0x00000000
A14     : 0x00000000  A15     : 0x0000000c  SAR     : 0x00000020  EXCCAUSE: 0x0000001c
EXCVADDR: 0x00000028  LBEG    : 0x40056f5c  LEND    : 0x40056f72  LCOUNT  : 0x00000000


Backtrace: 0x42008917:0x3fca23c0 0x42008934:0x3fca23f0 0x42011586:0x3fca2410 0x42010810:0x3fca2430 0x42024647:0x3fca2490 0x42027aa3:0x3fca24f0 0x420252a5:0x3fca2510 0x42024de5:0x3fca2530 0x4210c679:0x3fca2550 0x42027aa3:0x3fca2570 0x420252a5:0x3fca2590 0x4210cbc5:0x3fca25b0 0x42024ff9:0x3fca25d0 0x42027aa3:0x3fca25f0 0x4210cff1:0x3fca2610 0x42027aab:0x3fca2630 0x4210cff1:0x3fca2650 0x42027aab:0x3fca2670 0x42019ede:0x3fca2690 0x420056d4:0x3fca26e0 0x42008e68:0x3fca2700 0x42008f3a:0x3fca2770 0x42008f85:0x3fca27a0 0x42021e7b:0x3fca27c0 0x42021ed0:0x3fca2930 0x420160f2:0x3fca2950 0x4210a629:0x3fca2970 0x4201c9b5:0x3fca2990 0x42023872:0x3fca29b0 0x42023f3d:0x3fca29d0 0x420241f2:0x3fca29f0 0x4200ae6f:0x3fca2a30




ELF file SHA256: 2ffe29976

Rebooting...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4037e0ee
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce2820,len:0x1510
load:0x403c8700,len:0xccc
load:0x403cb700,len:0x2f48
entry 0x403c8914
I (24) boot: ESP-IDF 5.5.4 2nd stage bootloader
I (24) boot: Multicore bootloader
I (24) boot: chip revision: v0.2
I (24) boot: efuse block revision: v1.3
I (25) boot.esp32s3: Boot SPI Speed : 80MHz
I (25) boot.esp32s3: SPI Mode       : DIO
I (25) boot.esp32s3: SPI Flash Size : 16MB
I (25) boot: Enabling RNG early entropy source...
I (25) boot: Partition Table:
I (25) boot: ## Label            Usage          Type ST Offset   Length
I (26) boot:  0 otadata          OTA data         01 00 00009000 00002000
I (26) boot:  1 phy_init         RF data          01 01 0000b000 00001000
I (26) boot:  2 app0             OTA app          00 10 00010000 007c0000
I (27) boot:  3 app1             OTA app          00 11 007d0000 007c0000
I (27) boot:  4 nvs              WiFi data        01 02 00f90000 00070000
I (28) boot: End of partition table
I (28) esp_image: segment 0: paddr=007d0020 vaddr=3c130020 size=178c38h (1543224) map
I (302) esp_image: segment 1: paddr=00948c60 vaddr=3fc9bc00 size=04e88h ( 20104) load
I (306) esp_image: segment 2: paddr=0094daf0 vaddr=40378000 size=02528h (  9512) load
I (309) esp_image: segment 3: paddr=00950020 vaddr=42000020 size=121bech (1186796) map
I (519) esp_image: segment 4: paddr=00a71c14 vaddr=4037a528 size=116c8h ( 71368) load
I (535) esp_image: segment 5: paddr=00a832e4 vaddr=50000000 size=00020h (    32) load
I (543) boot: Loaded app from partition at offset 0x7d0000
I (543) boot: Disabling RNG early entropy source...
[I][logger:120]: Log initialized
[C][safe_mode:189]: Unsuccessful boot attempts: 0
[D][preferences:148]: Writing 1 items: 0 cached, 1 written, 0 failed
[I][app:060]: Running through setup()
```
Backtrace:
```
0x42008917: esphome::api::APIServer::on_light_update(esphome::light::LightState*) at /IDF_PROJECT/src/esphome/components/api/api_server.cpp:280
0x42008934: non-virtual thunk to esphome::api::APIServer::on_light_update(esphome::light::LightState*) at /IDF_PROJECT/src/esphome/components/api/api_server.h:90
0x42011586: esphome::ControllerRegistry::notify_light_update(esphome::light::LightState*) at /IDF_PROJECT/src/esphome/core/entity_types.h:35 (discriminator 3)
 (inlined by) esphome::light::LightState::publish_state() at /IDF_PROJECT/src/esphome/components/light/light_state.cpp:161 (discriminator 3)
0x42010810: esphome::light::LightCall::perform() at /IDF_PROJECT/src/esphome/components/light/light_call.cpp:175
0x42024647: esphome::light::LightControlAction<>::play() at /IDF_PROJECT/src/esphome/components/light/automation.h:54
0x42027aa3: esphome::Action<>::play_complex() at /IDF_PROJECT/src/esphome/core/automation.h:512
0x420252a5: esphome::ActionList<>::play() at /IDF_PROJECT/src/esphome/core/automation.h:597
 (inlined by) esphome::Automation<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:642
 (inlined by) esphome::Trigger<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:485
 (inlined by) esphome::script::SingleScript<>::execute() at /IDF_PROJECT/src/esphome/components/script/script.h:77
0x42024de5: setup()::{lambda()#28}::_FUN() at /home/trs/src/Voice/Devel/.esphome/packages/dc43710d/home-assistant-voice.yaml:1088
 (inlined by) _FUN at /home/trs/src/Voice/Devel/.esphome/packages/dc43710d/home-assistant-voice.yaml:1091
0x4210c679: esphome::StatelessLambdaAction<>::play() at /IDF_PROJECT/src/esphome/core/base_automation.h:242
0x42027aa3: esphome::Action<>::play_complex() at /IDF_PROJECT/src/esphome/core/automation.h:512
0x420252a5: esphome::ActionList<>::play() at /IDF_PROJECT/src/esphome/core/automation.h:597
 (inlined by) esphome::Automation<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:642
 (inlined by) esphome::Trigger<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:485
 (inlined by) esphome::script::SingleScript<>::execute() at /IDF_PROJECT/src/esphome/components/script/script.h:77
0x4210cbc5: void esphome::script::Script<>::execute_tuple_<>(std::tuple<> const&, std::integer_sequence<unsigned int>) at /IDF_PROJECT/src/esphome/components/script/script.h:57
 (inlined by) esphome::script::Script<>::execute_tuple(std::tuple<> const&) at /IDF_PROJECT/src/esphome/components/script/script.h:49
0x42024ff9: esphome::script::ScriptExecuteAction<esphome::script::Script<>>::play() at /IDF_PROJECT/src/esphome/components/script/script.h:218 (discriminator 1)
0x42027aa3: esphome::Action<>::play_complex() at /IDF_PROJECT/src/esphome/core/automation.h:512
0x4210cff1: esphome::Action<>::play_next_() at /IDF_PROJECT/src/esphome/core/automation.h:543
0x42027aab: esphome::Action<>::play_complex() at /IDF_PROJECT/src/esphome/core/automation.h:513
0x4210cff1: esphome::Action<>::play_next_() at /IDF_PROJECT/src/esphome/core/automation.h:543
0x42027aab: esphome::Action<>::play_complex() at /IDF_PROJECT/src/esphome/core/automation.h:513
0x42019ede: esphome::ActionList<>::play() at /IDF_PROJECT/src/esphome/core/automation.h:597
 (inlined by) esphome::Automation<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:642
 (inlined by) esphome::Trigger<>::trigger() at /IDF_PROJECT/src/esphome/core/automation.h:485
 (inlined by) esphome::voice_assistant::VoiceAssistant::client_subscription(esphome::api::APIConnection*, bool) at /IDF_PROJECT/src/esphome/components/voice_assistant/voice_assistant.cpp:537
0x420056d4: esphome::api::APIConnection::~APIConnection() at /IDF_PROJECT/src/esphome/components/api/api_connection.cpp:181
0x42008e68: std::default_delete<esphome::api::APIConnection>::operator()(esphome::api::APIConnection*) const at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/unique_ptr.h:93 (discriminator 1)
 (inlined by) std::default_delete<esphome::api::APIConnection>::operator()(esphome::api::APIConnection*) const at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/unique_ptr.h:87 (discriminator 1)
 (inlined by) std::__uniq_ptr_impl<esphome::api::APIConnection, std::default_delete<esphome::api::APIConnection> >::reset(esphome::api::APIConnection*) at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/unique_ptr.h:205 (discriminator 1)
 (inlined by) std::unique_ptr<esphome::api::APIConnection, std::default_delete<esphome::api::APIConnection> >::reset(esphome::api::APIConnection*) at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/unique_ptr.h:503 (discriminator 1)
 (inlined by) esphome::api::APIServer::remove_client_(unsigned char) at /IDF_PROJECT/src/esphome/components/api/api_server.cpp:188 (discriminator 1)
0x42008f3a: esphome::api::APIServer::loop() at /IDF_PROJECT/src/esphome/components/api/api_server.cpp:157
0x42008f85: esphome::api::APIServer::teardown() at /IDF_PROJECT/src/esphome/components/api/api_server.cpp:656
 (inlined by) esphome::api::APIServer::teardown() at /IDF_PROJECT/src/esphome/components/api/api_server.cpp:651
0x42021e7b: esphome::Application::teardown_components(unsigned long) at /IDF_PROJECT/src/esphome/core/application.cpp:356 (discriminator 1)
0x42021ed0: esphome::Application::safe_reboot() at /IDF_PROJECT/src/esphome/core/application.cpp:273
0x420160f2: esphome::restart::RestartButton::press_action() at /IDF_PROJECT/src/esphome/components/restart/button/restart_button.cpp:14
0x4210a629: esphome::button::Button::press() at /IDF_PROJECT/src/esphome/components/button/button.cpp:20
0x4201c9b5: std::_Function_handler<void (), esphome::web_server::WebServer::handle_button_request(esphome::web_server_idf::AsyncWebServerRequest*, esphome::web_server::UrlMatch const&)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at /IDF_PROJECT/src/esphome/components/web_server/web_server.cpp:796
 (inlined by) __invoke_impl<void, esphome::web_server::WebServer::handle_button_request(esphome::web_server_idf::AsyncWebServerRequest*, const esphome::web_server::UrlMatch&)::<lambda()>&> at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/invoke.h:61
 (inlined by) __invoke_r<void, esphome::web_server::WebServer::handle_button_request(esphome::web_server_idf::AsyncWebServerRequest*, const esphome::web_server::UrlMatch&)::<lambda()>&> at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/invoke.h:111
 (inlined by) _M_invoke at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/std_function.h:290
0x42023872: std::function<void ()>::operator()() const at /TOOLCHAIN/xtensa-esp-elf/include/c++/14.2.0/bits/std_function.h:591
 (inlined by) esphome::Scheduler::execute_item_(esphome::Scheduler::SchedulerItem*, unsigned long) at /IDF_PROJECT/src/esphome/core/scheduler.cpp:797
0x42023f3d: esphome::Scheduler::process_defer_queue_slow_path_(unsigned long&) at /IDF_PROJECT/src/esphome/core/scheduler.cpp:562
0x420241f2: esphome::Scheduler::process_defer_queue_(unsigned long&) at /IDF_PROJECT/src/esphome/core/scheduler.h:468
 (inlined by) esphome::Scheduler::call(unsigned long) at /IDF_PROJECT/src/esphome/core/scheduler.cpp:576
0x4200ae6f: esphome::Application::scheduler_tick_(unsigned long) at /IDF_PROJECT/src/esphome/core/application.h:569
 (inlined by) esphome::Application::loop() at /IDF_PROJECT/src/esphome/core/application.h:622
 (inlined by) esphome::loop_task(void*) at /IDF_PROJECT/src/esphome/components/esp32/core.cpp:25
```
Here are the logs when I modify `api_server.cpp` as indicated above:
```
[I][safe_mode:142]: Boot seems successful; resetting boot loop counter
[W][component:522]: safe_mode took a long time for an operation (88 ms), max is 30 ms
[D][preferences:148]: Writing 1 items: 0 cached, 1 written, 0 failed
[I][restart.button:011]: Restarting device
[I][app:271]: Rebooting safely
[D][debug:057]: Storing reboot source: web_server
[D][preferences:148]: Writing 2 items: 1 cached, 1 written, 0 failed
[D][api:285]: Client is NULL!
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x42008a85  PS      : 0x00060830  A0      : 0x82008aa3  A1      : 0x3fca23c0
A2      : 0x3fca72b0  A3      : 0x3fca6d34  A4      : 0x3c1457a8  A5      : 0x4201dfac
A6      : 0x3fca72d8  A7      : 0x3fca72d4  A8      : 0x82008a83  A9      : 0x3fca2360
A10     : 0x00000000  A11     : 0x3c13f8a7  A12     : 0x0000011d  A13     : 0x3c13091f
A14     : 0x00000000  A15     : 0x0000000c  SAR     : 0x0000001c  EXCCAUSE: 0x0000001c
EXCVADDR: 0x00000028  LBEG    : 0x40056f08  LEND    : 0x40056f12  LCOUNT  : 0x00000000


Backtrace: 0x42008a82:0x3fca23c0 0x42008aa0:0x3fca23f0 0x4201159e:0x3fca2410 0x42010828:0x3fca2430 0x4202465f:0x3fca2490 0x42027abb:0x3fca24f0 0x420252bd:0x3fca2510 0x42024dfd:0x3fca2530 0x4210c691:0x3fca2550 0x42027abb:0x3fca2570 0x420252bd:0x3fca2590 0x4210cbdd:0x3fca25b0 0x42025011:0x3fca25d0 0x42027abb:0x3fca25f0 0x4210d009:0x3fca2610 0x42027ac3:0x3fca2630 0x4210d009:0x3fca2650 0x42027ac3:0x3fca2670 0x42019ef6:0x3fca2690 0x420056d8:0x3fca26e0 0x42008e80:0x3fca2700 0x42008f52:0x3fca2770 0x42008f9d:0x3fca27a0 0x42021e93:0x3fca27c0 0x42021ee8:0x3fca2930 0x4201610a:0x3fca2950 0x4210a641:0x3fca2970 0x4201c9cd:0x3fca2990 0x4202388a:0x3fca29b0 0x42023f55:0x3fca29d0 0x4202420a:0x3fca29f0 0x4200ae87:0x3fca2a30




ELF file SHA256: 4a1937565

Rebooting...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4037e0ee
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce2820,len:0x1510
load:0x403c8700,len:0xccc
load:0x403cb700,len:0x2f48
entry 0x403c8914
I (24) boot: ESP-IDF 5.5.4 2nd stage bootloader
I (24) boot: Multicore bootloader
I (24) boot: chip revision: v0.2
I (24) boot: efuse block revision: v1.3
I (25) boot.esp32s3: Boot SPI Speed : 80MHz
I (25) boot.esp32s3: SPI Mode       : DIO
I (25) boot.esp32s3: SPI Flash Size : 16MB
I (25) boot: Enabling RNG early entropy source...
I (25) boot: Partition Table:
I (25) boot: ## Label            Usage          Type ST Offset   Length
I (26) boot:  0 otadata          OTA data         01 00 00009000 00002000
I (26) boot:  1 phy_init         RF data          01 01 0000b000 00001000
I (27) boot:  2 app0             OTA app          00 10 00010000 007c0000
I (27) boot:  3 app1             OTA app          00 11 007d0000 007c0000
I (27) boot:  4 nvs              WiFi data        01 02 00f90000 00070000
I (28) boot: End of partition table
I (28) esp_image: segment 0: paddr=00010020 vaddr=3c130020 size=178c48h (1543240) map
I (302) esp_image: segment 1: paddr=00188c70 vaddr=3fc9bc00 size=04e88h ( 20104) load
I (306) esp_image: segment 2: paddr=0018db00 vaddr=40378000 size=02518h (  9496) load
I (309) esp_image: segment 3: paddr=00190020 vaddr=42000020 size=121c04h (1186820) map
I (519) esp_image: segment 4: paddr=002b1c2c vaddr=4037a518 size=116d8h ( 71384) load
I (535) esp_image: segment 5: paddr=002c330c vaddr=50000000 size=00020h (    32) load
I (543) boot: Loaded app from partition at offset 0x10000
I (543) boot: Disabling RNG early entropy source...
[I][logger:120]: Log initialized
[C][safe_mode:189]: Unsuccessful boot attempts: 0
[D][preferences:148]: Writing 1 items: 0 cached, 1 written, 0 failed
[I][app:060]: Running through setup()
```

### Which version of ESPHome has the issue?

2026.5.2

### What type of installation are you using?

pip

### What platform are you using?

ESP32

### Component causing the issue

light

### YAML Config

```yaml
substitutions:
  name: home-assistant-voice-092088
  friendly_name: Home Assistant Voice
  ipv6_enable: "false"

packages:
  Nabu Casa.Home Assistant Voice PE: github://esphome/home-assistant-voice-pe/home-assistant-voice.yaml
  grove-i2c: github://esphome/home-assistant-voice-pe/modules/grove-i2c.yaml

esphome:
  name: ${name}
  name_add_mac_suffix: false
  friendly_name: ${friendly_name}

api:
  encryption:
    key: *REDACTED*

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

network:
  enable_ipv6: ${ipv6_enable}

web_server:
  port: 80
  include_internal: true

light:
  - id: !extend voice_assistant_leds
    internal: false
    name: "Voice Assistant Leds"
```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

The example YAML I have given is for a Voice Assistant Preview Edition device. However, I came across this issue when developing a custom component for an addressable light so the issue is not with the `esp32_rmt_led_strip` platform that the Voice Assistant uses.

In the YAML the web server is enabled to give access to the restart button and the light component is extended and named so that it appears in Home Assistant.

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Fixes a nullptr dereference crash in the API server during client teardown (issue #16798).

`APIServer::remove_client_` reset the client slot, which runs `~APIConnection()`, before decrementing `api_connection_count_`. `unique_ptr::reset()` nulls the slot before running the destructor, so during `~APIConnection()` the slot is already null but still inside the active range. On Home Assistant Voice PE, `~APIConnection()` unsubscribes from voice_assistant, whose `on_client_disconnected` automation publishes the LED light; that reenters `APIServer::on_light_update`, which iterates the active clients and dereferences the half-removed null slot, crashing the device on restart or disconnect.

The fix decrements the count before resetting the slot, so the slot being destroyed is excluded from the active range while `~APIConnection()` runs; reentrant state publishes during teardown then iterate only valid clients.

On the testing side; a host integration test was prototyped and confirmed the bug (it crashed before the fix and passed after), but reproducing the reentrancy on host requires publishing entity state from inside `~APIConnection()`, which is the only window where a client slot is null but still counted. That is not reachable on host without touching production code; voice_assistant (the real trigger) pulls in the ESP32 only audio stack, and the serial_proxy, zwave_proxy, and bluetooth_proxy unsubscribes in `~APIConnection()` are inert (they only null their stored connection pointer, they do not broadcast to active clients). The prototype seam and test are reverted in this branch, kept in history for reference, so production carries only the fix.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/esphome/issues/16798

**Pull request in [esphome.io](https://github.com/esphome/esphome.io) with documentation (if applicable):**

- esphome/esphome.io#N/A

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Reproduces in the field on Home Assistant Voice PE: voice_assistant
# unsubscribes during client teardown and its on_client_disconnected
# automation publishes a light, reentering the API state dispatch.
voice_assistant:
  microphone: mic
  on_client_disconnected:
    - light.toggle: led_ring
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome.io](https://github.com/esphome/esphome.io).

### CODE REVIEW

File: esphome/components/api/api_connection.cpp [hunk 1]
```diff
@@ -169,7 +169,16 @@ void APIConnection::start() {
   this->helper_->set_client_name(this->helper_->get_peername_to(peername), strlen(peername));
 }
 
+#ifdef USE_API_CONNECTION_TEST_HOOKS
+ConnectionDestroyTestHook api_connection_destroy_test_hook = nullptr;
+#endif
```
Comment (bdraco) [inline esphome/components/api/api_connection.cpp]: I spent many hours trying to build a repro test without a test hook, but the issue is so narrow and only affects voice pe , and given there is no way to mock voice pe in tests, this is what I landed on.

**Comment by bdraco** [comment]
Given this pr is a one line change and 187 lines of tests for a narrow edge case in voice pe and the mocking is ugly and has to be in production code, I'm going to strip it back out and ship the fix without a test because the test is worse that the cure

**Review by swoboda1337 — APPROVED** [review]

**Comment by bdraco** [comment]
thanks
